rfds/0001-signing-commits/text.md

```diff
@@ -0,0 +1,230 @@
+
+### Compressing signed commits
+
+Another problem with signing commits is that signatures do not compress well. We have done a great deal of work to compress away the hashes on each commit in an Automerge document, retaining the signatures for each commit would undo all that work. We need to know who is attesting to each commit though, what can we do?
+
+Well, we can perform the same kind of trick that we do with the change hashes. The insight is that in linear runs of commits we only need the signature of the end of the run. E.g. in this graph where hexagons are signed by Bob, rhombuses by Alice, and circles by Charlie.
+
+```mermaid
+graph TD
+A{{A}}
+B{{B}}
+C((C))
+D((D))
+E{E}
+F{F}
+A --> B
+B --> C
+C --> D
+A --> E
+E --> F
+```
+
+Conceptually we only need to store the last signature in each run, plus a counter saying how many changes back the run goes. So we would store something like:
+
+| Head of run | Count |
+| ----------- | ----- |
+| D           | 2     |
+| F           | 2     |
+| B           | 2     |
+
+All well and good, but what are we actually signing at the end of each run? If the signature is just over the commit (its hash and its parent hashes) then when we receive a run of commits we have no way of verifying them. Say we had this run of commits originally:
+
+```mermaid
+graph LR
+A["`A
+signed by bob`"]
+B["`B
+signed by charlie`"]
+C["`C
+signed by bob`"]
+A --> B
+B --> C
+```
+
+But Bob sends you a compressed chunk containing A, B, and C, Bob can just send you the signature on C and tell you the run was three commits long and you will infer that Bob was the author of B!
+
+The problem is the same as with the multiple signatures on a commit problem, we have no way of binding the contents of a commit to an author. In this case we can solve the problem by requiring that signatures not be over the commit directly, but over a wrapper we will call an attestation:
+
+```rust
+struct AttestationPayload<CommitId> {
+	commit: CommitId,
+	parent_attestation: AttestationHash,
+	author: ed25519_dalek::VerifyingKey
+}
+
+struct Attestation<CommitId> {
+	payload: AttestationPayload<CommitId>,
+	signature: ed25519_dalek::Signature,
+}
+
+// The hash of a n`AttestationPayload
+struct AttestationHash(blake3::Hash)
+```
+
+Now, in order to verify the run we can start at head of the run and walk backwards `count` steps, then we can construct the `AttestationPayload` for each commit in the run, we can check that the hash of the penultimate `AttestationPayload` matches the `parent_attestation` in the `Attestation` for the head of the run before validating the signature.
+
+## Attestations
```

Comment:
I think the version in my earlier comment can be seen as a simplification of this approach. Lemme know if I'm missing something!